### PR TITLE
[js/webgpu] supported depthToSpace operator in webgpu

### PR DESCRIPTION
**Pull request by MasayoshiTsutsui**
### Description
This PR supports [DepthToSpace](https://onnx.ai/onnx/operators/onnx__DepthToSpace.html#depthtospace) operator in webgpu backend.

### Test
We followed the steps described on [this page](https://gist.github.com/fs-eire/a55b2c7e10a6864b9602c279b8b75dce) to build, tested with the following commands, and confirmed that it passed the Model and Op tests that already existed. (Probably, these test cases were prepared in the past for WebGL backend)
```
~/onnxruntime/js/web>
% npm test -- suite0 -b=webgpu --wasm-number-threads=1 --debug   
```
##### NOTE
I want to tell you that the main branch version failed 5 tests for the resize_upsample_sizes_nearest operator.
Since I didn't touch this issue, those test cases still fail in my branch as well.
Should I post an issue for this?

### Motivation and Context
Though the DepthToSpace operator plays a crucial role in super-resolution domains, it was not supported in webgpu backend.

### CODE REVIEW

**Comment by MasayoshiTsutsui** [comment]
@microsoft-github-policy-service agree [company="{your company}"]

**Comment by MasayoshiTsutsui** [comment]
@microsoft-github-policy-service agree

**Comment by MasayoshiTsutsui** [comment]
@fs-eire 
I'm sorry for mentioning you roughly.
I couldn't find how to assign reviewers in this PR.
It is probably because this is my first time contributing to this project, so I may not be allowed to assign reviewers.
Could you teach me what I should do?

**Comment by fs-eire** [comment]
> @fs-eire I'm sorry for mentioning you roughly. I couldn't find how to assign reviewers in this PR. It is probably because this is my first time contributing to this project, so I may not be allowed to assign reviewers. Could you teach me what I should do?

Hi @MasayoshiTsutsui thank you for the contribution! Don't worry about the reviewer assignment - we will surely take a look at the change.

**Comment by MasayoshiTsutsui** [comment]
```
/mnt/vss/_work/1/s/js/web/lib/wasm/jsep/webgpu/ops/depth-to-space.ts
  40:1  error  This line has a length of 123. Maximum allowed is 120                                             max-len
  60:9  error  Identifier 'shape_before_perm' is not in camel case                                               camelcase
  60:9  error  Variable name `shape_before_perm` must match one of the following formats: camelCase, UPPER_CASE  @typescript-eslint/naming-convention
  61:9  error  Identifier 'shape_after_perm' is not in camel case                                                camelcase
  61:9  error  Variable name `shape_after_perm` must match one of the following formats: camelCase, UPPER_CASE   @typescript-eslint/naming-convention
```

It seems my code style doesn't follow the coding rule.  I will revise my codes immediately.
Is there a way to run such a CI-checking script locally? If there exists, I would appreciate it if you could tell me which document teaches it to me.

**Comment by gyagp** [comment]
> ```
> /mnt/vss/_work/1/s/js/web/lib/wasm/jsep/webgpu/ops/depth-to-space.ts
>   40:1  error  This line has a length of 123. Maximum allowed is 120                                             max-len
>   60:9  error  Identifier 'shape_before_perm' is not in camel case                                               camelcase
>   60:9  error  Variable name `shape_before_perm` must match one of the following formats: camelCase, UPPER_CASE  @typescript-eslint/naming-convention
>   61:9  error  Identifier 'shape_after_perm' is not in camel case                                                camelcase
>   61:9  error  Variable name `shape_after_perm` must match one of the following formats: camelCase, UPPER_CASE   @typescript-eslint/naming-convention
> ```
> 
> It seems my code style doesn't follow the coding rule. I will revise my codes immediately. Is there a way to run such a CI-checking script locally? If there exists, I would appreciate it if you could tell me which document teaches it to me.

You may run "npm run lint" in /js folder.

**Comment by MasayoshiTsutsui** [comment]
@fs-eire @gyagp 
Thank you for your reply!
I've modified the codes and checked that the above linter errors are resolved.
Could you review it again?

p.s.
I would like you to notice what I noted in the first comment: the existing issues around the resize_upsample_sizes_nearest operator.

> I want to tell you that the main branch version failed 5 tests for the resize_upsample_sizes_nearest operator.
> Since I didn't touch this issue, those test cases still fail in my branch as well.
> Should I post an issue for this?

**Comment by MasayoshiTsutsui** [comment]
I've resolved errors and warnings from lintrunner and formatter.
But I ignored the following 2 kinds of warnings from reviewdog in favor of following the style of other files.

- Include the directory when naming header files  [build/include_subdir] [4]
- Lines should be <= 120 characters long  [whitespace/line_length] [2]

File: js/web/lib/wasm/jsep/webgpu/ops/depth-to-space.ts [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    perm = isDCRmode ? [0, 3, 4, 1, 5, 2] : [0, 1, 4, 2, 5, 3];
+  }
+  const reshapedInputTensor = inputTensor.reshape(shape);
+
+  const shapeBeforePerm = reshapedInputTensor.dims;
+  const shapeAfterPerm = ShapeUtil.sortBasedOnPerm(shapeBeforePerm, perm);
+
+  const inputDataType = inputTensor.dataType;
+  const reshapedInput = inputVariable('a', inputDataType, shapeBeforePerm);
+  const permedOutput = outputVariable('output', inputDataType, shapeAfterPerm);
```
Comment (fs-eire) [inline js/web/lib/wasm/jsep/webgpu/ops/depth-to-space.ts:66]: There is a mismatch between the usage of `inputVariable`/`outputVariable` functions and the uniform data.

There are 2 ways using the 3rd parameter of `inputVariable`/`outputVariable` functions:

- using a shape (number[]), so the shape info will be hardcoded in the source
- using a rank (number), so the shape info is expected to be included in the `programUniforms` property returned by `getRunData()` function.

I assume that you want to use uniform, so changing the shape to rank here can fix this problem.

File: js/web/lib/wasm/jsep/webgpu/ops/depth-to-space.ts [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    shaderCache: {hint: `${inputTensor.dims};${attributes.blocksize};${attributes.mode}`, inputDependencies: ['rank']},
+    getRunData: (inputs) => {
+      const outputShape = isChannelLast ? [n, h * blocksize, w * blocksize, c / (blocksize ** 2)] :
+                                          [n, c / (blocksize ** 2), h * blocksize, w * blocksize];
+      const outputSize = ShapeUtil.size(outputShape);
+      return {
+        outputs: [{dims: outputShape, dataType: inputs[0].dataType}],
+        dispatchGroup: {x: Math.ceil(outputSize / 64 /* workgroup size */)},
+        programUniforms:
+            [{type: DataType.uint32, data: outputSize}, ...createTensorShapeVariables(inputs[0].dims, outputShape)],
```
Comment (fs-eire) [inline js/web/lib/wasm/jsep/webgpu/ops/depth-to-space.ts:94]: `createTensorShapeVariables()` should use logical shape (ie. shapeBeforePerm and shapeAfterPerm)

**Comment by fs-eire** [comment]
> @fs-eire @gyagp Thank you for your reply! I've modified the codes and checked that the above linter errors are resolved. Could you review it again?
> 
> p.s. I would like you to notice what I noted in the first comment: the existing issues around the resize_upsample_sizes_nearest operator.
> 
> > I want to tell you that the main branch version failed 5 tests for the resize_upsample_sizes_nearest operator.
> > Since I didn't touch this issue, those test cases still fail in my branch as well.
> > Should I post an issue for this?

resize_upsample_sizes_nearest does not fail on my Windows desktop. Could you share the OS version, browser version and GPU of your development environment?

File: onnxruntime/core/providers/js/operators/depth_to_space.h [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+// Licensed under the MIT License.
+
+#pragma once
+
+#include "core/providers/js/js_kernel.h"
+
+namespace onnxruntime {
+namespace js {
+
+template <bool has_mode, bool is_channels_last>
```
[the remark this answers is not in the document]
Comment (MasayoshiTsutsui) [reply onnxruntime/core/providers/js/operators/depth_to_space.h]: I forgot to erase them in the debugging phase. I've removed it in the additional commit.

File: onnxruntime/core/providers/js/operators/depth_to_space.h [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+                                 "mode" : UTF8ToString($2),
+                                 "format" : $3 ? "NHWC" : "NCHW"
+                               }),
+                               static_cast<int32_t>(blocksize_),
+                               mode_.c_str(), static_cast<int32_t>(is_channels_last));
+  }
+
+ private:
+  int64_t blocksize_;
+  std::string mode_;
```
[the remark this answers is not in the document]
Comment (MasayoshiTsutsui) [reply onnxruntime/core/providers/js/operators/depth_to_space.h]: Modified the codes to treat them as local variables in the additional commit.

File: js/web/lib/wasm/jsep/webgpu/ops/depth-to-space.ts (hunk 1 above)
Comment (MasayoshiTsutsui) [reply js/web/lib/wasm/jsep/webgpu/ops/depth-to-space.ts]: I've understood the programUniforms functionality thanks to your explanations.
I changed the shape to rank in the additional commit.

File: js/web/lib/wasm/jsep/webgpu/ops/depth-to-space.ts (hunk 2 above)
Comment (MasayoshiTsutsui) [reply js/web/lib/wasm/jsep/webgpu/ops/depth-to-space.ts]: Use logical shape in the additional commit.

**Comment by MasayoshiTsutsui** [comment]
My  develop environment is as follows.

- MacOS Ventura 13.4.1
- Chrome 123.0.6312.59（Official Build） （arm64）
- GPU: Apple M1 pro

I provide one of the error messages below.
The test fails in opset11, 13, 17, 18 and 19.

>   #ModelTest# - node-opset11-webgpu
>     [webgpu] test_resize_upsample_sizes_nearest
>       ✖ test_data_set_0
>         Chrome 123.0.0.0 (Mac OS 10.15.7)
>       AssertionError: tensor data should match: expected false to be true
>           at _TensorResultValidator.checkApiTensorResult (test/ort.test.js:41906:60)
>           at _TensorResultValidator.checkNamedTensorResult (test/ort.test.js:41914:16)
>           at runModelTestSet (test/ort.test.js:41642:18)
>           at async Context.<anonymous> (test/ort.test.js:44282:15)

**Review by fs-eire — APPROVED** [review]

**Pull request merged**